[FLINK-37834] Flink JDBC support ClickHouse Dialect

Biasqo opened the pull request:
🔌 ClickHouse Dialect for Flink Link JDBC Connector

This project provides a ClickHouse SQL dialect implementation for the Apache Flink JDBC Connector.

🚀 Purpose
Apache Flink’s Link JDBC connector offers a flexible way to integrate with relational databases through dialects. However, native support for ClickHouse’s SQL features (e.g. merge trees, upserts, batch writes) is limited in SQL CLI.

✨ Features
✅ Custom ClickHouseDialect integration
✅ Column quoting and identifier case sensitivity fixes
✅ Insert/update query support tailored for ClickHouse (pkFields removed from update statement)

TODO:
Refactor Tests: Some methods are private (like getStreamFields() that's why I had to recreate some test classes)

Jira Issue:
https://issues.apache.org/jira/projects/FLINK/issues/FLINK-37834

eskabetxe commented on .idea/vcs.xml:
```diff
@@ -1,24 +1,24 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
```
the changes in this file should be reverted

eskabetxe commented on flink-connector-jdbc-clickhouse/pom.xml:
```diff
@@ -0,0 +1,94 @@
+    <packaging>jar</packaging>
+
+    <properties>
+        <clickhouse.version>0.6.2</clickhouse.version>
```
Why not use the last version available 0.8.6?

eskabetxe reviewed (changes requested):
Thanks for your contribution, I can't start CI/CD sorry. @1996fanrui can you start it?

I review the main code and left minor comments.
I dont review the tests as you have a TODO to some refactoring there.

You also need to add this module on flink-connector-jdbc-architecture.

Biasqo commented:
> Thanks for your contribution, I can't start CI/CD sorry. @1996fanrui can you start it?
> 
> 
> 
> I review the main code and left minor comments.
> 
> I dont review the tests as you have a TODO to some refactoring there.
> 
> 
> 
> You also need to add this module on flink-connector-jdbc-architecture.

Is it okay if I refactor some of the core classes in the tests, like making them protected?

eskabetxe commented:
> @eskabetxe Thanks for getting back to me
> 
> 1. I’m using version 0.6.2 since 0.8+ has some dependency issues with httpcore, and I didn’t want to change the existing httpcore version. If it’s okay to upgrade some of the libraries already included in the distribution, I can go ahead and fix it.
> 2. I will revert .vcs
> 3. About the tests: I’m not sure if I can override some private methods that are part of abstract test classes (for example, the abstract class TableBase, method getStreamFields). If I’m allowed to make them protected, the number of test classes could be significantly reduced. Clickhouse has its unique create statement - that's why it was a problem

Its ok to bump some versions and is ok also change some test methods from private to protected if you need to override them..

Biasqo committed:
Adding clickhouse dialect 1. Refactored tests. 2. Added clickhouse to archunit, 3. Driver version for tests = 0.6.2 (ClickHouse testcontainers support that version) 4. Refactored UPDATE statement

Biasqo committed:
Mini fix for methods accessibility

Biasqo commented:
@davidradl maybe you can review?

Biasqo committed:
Merge branch 'apache:main' into main

Biasqo committed:
1. refactored some tests and CI fixes

Biasqo committed:
Merge pull request #1 from Biasqo/prerelease

1. refactored some tests and CI fixes

och5351 commented:
Hi, @Biasqo !
Are you still looking at this PR?

Biasqo commented:
Hi, @och5351 
The code was ready for its time.
I've waited for so long and got no answer.
I don't know if anyone actually needs a clickhouse connector, so it's up to you.

och5351 commented:
@Biasqo 
I think the ClickHouse connector will be really useful.
I will review and support this PR as soon as I have time (as soon as possible).
Of course, as long as you're okay with it.

Thank you for your contribution.

Biasqo commented:
@och5351 
Alright,
If there are any issues with the pr leave a comment and I'll do my best to fix them.

och5351 commented on flink-connector-jdbc-clickhouse/pom.xml:
```diff
@@ -0,0 +1,95 @@
+    <packaging>jar</packaging>
+
+    <properties>
+        <clickhouse.jdbc.version>0.6.2</clickhouse.jdbc.version>
```
It seems not dependency(commons-io) about version. so I think It's working 0.9.7 :) 
please let me know if i'm wrong.

Biasqo replied:
Done)

och5351 commented on flink-connector-jdbc-clickhouse/src/test/java/org/apache/flink/connector/jdbc/clickhouse/testutils/ClickHouseImages.java:
```diff
@@ -0,0 +1,23 @@
+
+/** ClickHouse docker images. */
+public interface ClickHouseImages {
+    String CLICKHOUSE_LATEST = "clickhouse/clickhouse-server:24.3";
```
Could you please version up 25.8? (It seems the current version isn't supported.)
My test environmental is that version. sorry.

Biasqo replied:
Changed to 25.8

och5351 reviewed (commented):
I am writing this review to continue striving for greater perfection.

Please feel free to point out any errors without reservation.

Biasqo committed:
Merge branch 'apache:main' into main

Biasqo committed:
Biasqo:develop to Biasqo:main

1. Changed deprecated com.clickhouse.jdbc.ClickHouseDriver to com.clickhouse.jdbc.Driver, 2. Added commons-io (ver 2.20.0 to be compatible with clickhouse jdbc 0.9.7), 3. Updated clickhouse driver from 0.6.2 to 0.9.7, 4. Added clickhouse lineage (consider adding it into openlineage)

Biasqo commented:
@och5351 
I've made some changes you mentioned above
1. I've updated clickhouse server to 25.8
2. I've updated clickhouse driver to 0.9.7
3. I've added lineage to clickhouse (I think its better to add that clickhouse support into openlineage)
4.  I've also changed deprecated class com.clickhouse.jdbc.ClickHouseDriver to com.clickhouse.jdbc.Driver class

I've passed the tests in my forked repo, so its ready to merge after your review

och5351 commented on flink-connector-jdbc-clickhouse/src/main/java/org/apache/flink/connector/jdbc/clickhouse/database/ClickHouseFactory.java:
```diff
@@ -0,0 +1,50 @@
+public class ClickHouseFactory implements JdbcFactory {
+    @Override
+    public boolean acceptsURL(String url) {
+        return url.startsWith("jdbc:clickhouse:");
```
Should we support the jdbc:ch: alias as well?

```suggestion
return url.startsWith("jdbc:clickhouse:") || url.startsWith("jdbc:ch:");
```

Biasqo replied:
That's a good idea
I will add it

och5351 commented on flink-connector-jdbc-clickhouse/src/main/java/org/apache/flink/connector/jdbc/clickhouse/database/lineage/ClickHouseJdbcExtractor.java:
```diff
@@ -0,0 +1,43 @@
+    public JdbcLocation extract(String rawUri, Properties properties) throws URISyntaxException {
+        return this.delegate().extract(rawUri, properties);
+    }
+}
```
This is non-blocking issue.

The default port could be 8123 per the https://clickhouse.com/docs/integrations/language-clients/java/jdbc#configuration. 

Also, would you consider supporting the jdbc:ch: alias by normalizing it to clickhouse:?

```suggestion
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.connector.jdbc.clickhouse.database.lineage;

import io.openlineage.client.utils.jdbc.JdbcExtractor;
import io.openlineage.client.utils.jdbc.JdbcLocation;
import io.openlineage.client.utils.jdbc.OverridingJdbcExtractor;

import java.net.URISyntaxException;
import java.util.Properties;

/** Implementation of {@link JdbcExtractor} for ClickHouse. */
public class ClickHouseJdbcExtractor implements JdbcExtractor {

    public ClickHouseJdbcExtractor() {}

    private JdbcExtractor delegate() {
        return new OverridingJdbcExtractor("clickhouse", "8123");
    }

    @Override
    public boolean isDefinedAt(String jdbcUri) {
        return jdbcUri.startsWith("ch:") || jdbcUri.startsWith("clickhouse:");
    }

    @Override
    public JdbcLocation extract(String rawUri, Properties properties) throws URISyntaxException {
    // Normalize scheme and remove protocol prefix (similar to how PostgreSQL removes SSL info)
    String normalizedUri =
        rawUri
            .replaceFirst("^ch:", "clickhouse:")
            .replaceFirst("^clickhouse:(https?|grpc)://", "clickhouse://");
        return delegate().extract(normalizedUri, properties);
    }
}
```

Biasqo replied:
I will change the port but in my experience ssl is always turned on, so 8443 port becomes default)

och5351 replied:
Actually, I think so too. However, the documentation seems to list HTTP as the default.

<img width="449" height="30" alt="image" src="https://github.com/user-attachments/assets/83285486-5ae9-43c7-9b98-d106da2174cf" />


In fact, if this phrase hadn't been there, I think your opinion would have been more correct.

och5351 reviewed (changes requested):
Thanks @Biasqo!
I've submitted an OpenLineage PR for this (https://github.com/OpenLineage/OpenLineage/pull/4495), but it will take time to merge.

@RocMarshal  @eskabetxe Should we use this custom extractor for now and migrate later, or would you prefer to wait for the OpenLineage release?

Biasqo committed:
1. Minor fixes requested by och5351: added ch prefix to ClickHouseFactory and JdbcExtractor, 2. Refactored ClickHouseJdbcExtractor (#4)

Biasqo commented:
@och5351 
Thank you for your review
Added ch prefix + suggested your changes in latest commit

Upd:
Should i also add some documentation on clickhouse connector?
I've seen some documentation in docs folder

eskabetxe commented on flink-connector-jdbc-clickhouse/src/test/java/org/apache/flink/connector/jdbc/clickhouse/table/ClickHouseDynamicTableSourceITCase.java:
```diff
@@ -0,0 +1,75 @@
+
+    @Override
+    public String timestampFilterExpression() {
+        return "timestamp6_col > TO_TIMESTAMP_LTZ('2020-01-01 15:35:00') "
```
Why we change this to a zoned timestamp??

Biasqo replied:
That's an old typo
I will refresh the tests, there are not enough tests for clickhouse

davidradl reviewed (commented):
sorry updated by mistake

och5351 commented:
Hi @RocMarshal, @davidradl, @eskabetxe, @Biasqo!

Could you please share your thoughts on this plan?

> Flink Clickhouse JDBC Connector needs OpenLineage support.
1. Wait for the PR in OpenLineage to be merged and update the version in this repo. (https://github.com/OpenLineage/OpenLineage/pull/4495)
2. First merge the custom lineage implementation, then update the version and apply it.

That's the plan I came up with. If you have any better ideas or advice, 
I'd really appreciate it if you could share them!

Biasqo committed:
ClickHouse dialect: Test refactoring

1. Refactored ClickHouseDynamicTableSourceITCase (adding more columns to source table)
2. Added MAP type support to JdbcTypeUtil (since MAP type is actively used in ClickHouse DB)
3. Refactored ClickHouseDialectConverter: added MAP and ARRAY support to internal converter and did some minor fixes

Biasqo commented:
 @eskabetxe @och5351 
I refactored  ClickHouse tests and they look cleaner now.
Also fixed minor bugs in ClickHouse converter.

Biasqo commented:
> Hi @RocMarshal, @davidradl, @eskabetxe, @Biasqo!
> 
> Could you please share your thoughts on this plan?
> 
> > Flink Clickhouse JDBC Connector needs OpenLineage support.
> 
> 1. Wait for the PR in OpenLineage to be merged and update the version in this repo. ([flink: add ClickHouse JDBC support OpenLineage/OpenLineage#4495](https://github.com/OpenLineage/OpenLineage/pull/4495))
> 2. First merge the custom lineage implementation, then update the version and apply it.
> 
> That's the plan I came up with. If you have any better ideas or advice, I'd really appreciate it if you could share them!

I think we should wait for the PR in OpenLineage, it will keep the codebase clean

Biasqo committed:
Merge branch 'apache:main' into main

Biasqo committed:
Updating Openlineage version (1.47.1) to support ClickHouse DB

* Adding openlineage for ClickHouse

* Minor compatibility tweaks

Biasqo commented:
Hi @och5351, @davidradl, @eskabetxe!
Ive added OpenLineage native support for ClickHouse DB and made some minor fixes. 
Looking forward to seeing your comments!

Biasqo commented:
Hi @och5351, @davidradl, @eskabetxe!
Could you please rerun checks? 
The error seems to be in core module, I've passed all checks in my forked main branch

Biasqo commented:
> Hi @och5351, @davidradl, @eskabetxe!
> 
> Could you please rerun checks? 
> 
> The error seems to be in core module, I've passed all checks in my forked main branch

CI seems to be unstable once again
Now the error is in Db2 module

och5351 commented:
By the way, just to confirm — was this Jira ticket 
(FLINK-37834) created by @Biasqo, and is the account name Arsene?

> https://issues.apache.org/jira/browse/FLINK-37834

If that's correct, would it be okay to ask @RocMarshal 
to assign it?

och5351 commented on flink-connector-jdbc-clickhouse/src/main/java/org/apache/flink/connector/jdbc/clickhouse/database/dialect/ClickHouseDialectConverter.java:
```diff
@@ -0,0 +1,271 @@
+                        "BINARY/VARBINARY types are not supported by ClickHouse dialect. "
+                                + "Use STRING instead.");
+            case CHAR:
+            case VARCHAR:
```
This is already handled identically in AbstractDialectConverter. Consider removing this case and delegating via a default fallthrough to avoid duplication.

och5351 commented on flink-connector-jdbc-clickhouse/src/main/java/org/apache/flink/connector/jdbc/clickhouse/database/dialect/ClickHouseDialectConverter.java:
```diff
@@ -0,0 +1,271 @@
+        switch (type.getTypeRoot()) {
+            case NULL:
+                return null;
+            case BOOLEAN:
```
This is already handled identically in AbstractDialectConverter. Consider removing this case and delegating via a default fallthrough to avoid duplication.

och5351 commented on flink-connector-jdbc-clickhouse/src/main/java/org/apache/flink/connector/jdbc/clickhouse/database/dialect/ClickHouseDialectConverter.java:
```diff
@@ -0,0 +1,271 @@
+    @Override
+    public JdbcDeserializationConverter createInternalConverter(LogicalType type) {
+        switch (type.getTypeRoot()) {
+            case NULL:
```
This is already handled identically in AbstractDialectConverter. Consider removing this case and delegating via a default fallthrough to avoid duplication.

och5351 commented on flink-connector-jdbc-clickhouse/src/main/java/org/apache/flink/connector/jdbc/clickhouse/database/dialect/ClickHouseDialectConverter.java:
```diff
@@ -0,0 +1,271 @@
+                                : val instanceof Timestamp
+                                        ? TimestampData.fromTimestamp((Timestamp) val)
+                                        : val;
+            case DECIMAL:
```
This is already handled identically in AbstractDialectConverter. Consider removing this case and delegating via a default fallthrough to avoid duplication.

Biasqo commented:
> By the way, just to confirm — was this Jira ticket ([FLINK-37834](https://issues.apache.org/jira/browse/FLINK-37834)) created by @Biasqo, and is the account name Arsene?
> 
> > https://issues.apache.org/jira/browse/FLINK-37834
> 
> If that's correct, would it be okay to ask @RocMarshal to assign it?

Yes thats correct, I created FLINK-37834, Arsene is my name)))
Thx for checking

och5351 commented on flink-connector-jdbc-clickhouse/src/main/java/org/apache/flink/connector/jdbc/clickhouse/database/dialect/ClickHouseDialectConverter.java:
```diff
@@ -0,0 +1,271 @@
+    }
+
+    // adding support to MAP and ARRAY types
+    public static Object toExternalSerializer(Object value, LogicalType type) {
```
How about changing it to private?

och5351 commented on flink-connector-jdbc-clickhouse/src/main/java/org/apache/flink/connector/jdbc/clickhouse/database/dialect/ClickHouseDialectConverter.java:
```diff
@@ -0,0 +1,271 @@
+                                                () ->
+                                                        new RuntimeException(
+                                                                "Unknown array element type"));
+                final JdbcDeserializationConverter elementConverter =
```
Consider using `createNullableInternalConverter` instead of `createInternalConverter` for the element converter. `PostgresDialectConverter` follows this pattern, which delegates null handling to the framework and removes the need for the manual null check on the next line.

och5351 commented on flink-connector-jdbc-clickhouse/src/main/java/org/apache/flink/connector/jdbc/clickhouse/database/dialect/ClickHouseDialectConverter.java:
```diff
@@ -0,0 +1,271 @@
+                    }
+                    return new GenericArrayData(converted);
+                };
+            case MAP:
```
It would be cleaner to extract the `ARRAY` and `MAP` logic into separate private methods (e.g., createArrayConverter, createMapConverter), following the pattern in `PostgresDialectConverter.createPostgresArrayConverter`. 

This keeps the switch cases concise and separates traversal logic from null handling.
What do you think about handling it this way?

och5351 commented:
Hi, @Biasqo !
I apologize for the delay.

PTAL !

och5351 commented on flink-connector-jdbc-clickhouse/src/main/java/org/apache/flink/connector/jdbc/clickhouse/database/dialect/ClickHouseDialectConverter.java:
```diff
@@ -0,0 +1,271 @@
+            case TIME_WITHOUT_TIME_ZONE:
+                return Time.valueOf(LocalTime.ofNanoOfDay((int) value * 1_000_000L));
+            case TIMESTAMP_WITH_TIME_ZONE:
+            case TIMESTAMP_WITHOUT_TIME_ZONE:
```
```suggestion
            case TIMESTAMP_WITH_TIME_ZONE:
            case TIMESTAMP_WITHOUT_TIME_ZONE:
                   return ((TimestampData) value).toTimestamp();
```

The value here is `TimestampData` (Flink's internal type returned by ElementGetter), not `LocalDateTime` or `Timestamp`. This falls through to : value, returning `TimestampData` directly to the ClickHouse JDBC driver, which will cause a failure. Following the same approach as `AbstractDialectConverter.createExternalConverter`, which calls val.getTimestamp(...).toTimestamp()

Biasqo replied:
Should I do same for DATE type?

och5351 replied:
No, DATE already returns java.sql.Date correctly.

och5351 commented on flink-connector-jdbc-clickhouse/src/main/java/org/apache/flink/connector/jdbc/clickhouse/database/dialect/ClickHouseDialectConverter.java:
```diff
@@ -0,0 +1,271 @@
+                        : value instanceof Timestamp
+                                ? TimestampData.fromTimestamp((Timestamp) value)
+                                : value;
+            case DECIMAL:
```
```suggestion
            case DECIMAL:
                   return ((DecimalData) value).toBigDecimal();
```

Biasqo committed:
Merge branch 'apache:main' into main

Biasqo committed:
Hotfix/FLINK-37834

Refactored code based on requested changes

Biasqo commented:
Hi @och5351!
Ive refactored converter based on your changes and added MAP and ARRAY into tests (latest commit [a1d4bfc](https://github.com/apache/flink-connector-jdbc/pull/171/commits/a1d4bfc49de2235d56a68cc635b51486771a8c14))
Let me know if there are any remaining comments.

Biasqo committed:
Minor fix for MAP type

Biasqo committed:
ClickHouse sink and source tests: Adding MAP and ARRAY data

* Added test for MAP and ARRAY in JdbcSink and JdbcSource

Biasqo commented:
CI checks are failing, but they're all passing on my branch.
<img width="1271" height="598" alt="ci_checks" src="https://github.com/user-attachments/assets/d1411c43-ca9c-4678-90eb-aff262b8a602" />

RocMarshal commented:
Thanks @Biasqo @och5351 

Could you @eskabetxe help take a look ? thx
